File#truncate

porras opened the pull request:
This PR implements `File#truncate`, a direct mapping to the `ftruncate` system call, and equivalent to [Ruby's method](http://www.rubydoc.info/stdlib/core/File#truncate-instance_method). The spec shows a use case that comes to mind: opening a file with `r+`, reading it complete, and then overwriting it completely (if the new content is shorter than the old, we need to truncate).

jhass commented on src/file.cr:
```diff
@@ -377,6 +378,15 @@ class File < IO::FileDescriptor
   end
 
+  # Truncates the file to the specified length. Requires a write file descriptor
+  def truncate(length)
```
I think a default of `0` makes sense here.

porras replied:
A default makes total sense. I was thinking about another possibility, let me know what you think: the current `pos`. This way you can stick to the default for the two main (that I can think of) use cases:
* You just opened the file and want to truncate it entirely (current `pos` would be `0`)
* You just finished writing and want to truncate the rest (current `pos` would be where you finished writing)
I think this makes sense but it might have problems I'm not seeing, if you don't like it, I'll just add the `0` default, otherwise I'll implement the one I just described.

Thanks for the feedback!

jhass replied:
I think that might be a bit hard to understand and unexpected for the default behaviour.  However I see its usefulness, how about doing that when passing `nil` or `true` for `length` (which should be called `size` btw)?

porras replied:
mmm I have been playing with this and I think I don't understand well how the argument passed to the system call works. For example, this test passes, when I think it should fail (it should leave an empty file, I'd expect):

```crystal
      File.write(filename, "0123456789")
      File.open(filename, "r+") do |f|
        f.read.should eq("0123456789")
        f.rewind
        f.puts("333")
        f.truncate(0)
      end

      File.read(filename).should eq("333\n")
      File.size(filename).should eq(4)
```

Maybe it's interpreted *relative to the current position*? Apparently not, because if I try to pass a negative argument, it fails:

```
Error truncating file '/Users/sergio/Code/crystal/spec/std/data/temp_write.txt': Invalid argument
```

Sorry, I have never done file manipulation directly with system calls so I'm confused :disappointed:

jhass replied:
You're interpreting the manpage right, the issue is the buffered IO here. `truncate` should flush first, else you truncate the file and then it's flushed at close -> stuff pending in the write buffer is written.

jhass replied:
@asterite we were discussing an issue in this outdated comment thread here, sorry :P

porras replied:
Yes, sorry I didn't follow up, I've had a couple of very busy days (and tomorrow will be as well). I was playing with specifying a size and flushing the buffer and I'm still a bit confused by the behavior, I've seen some inconsistent results like the one I mentioned (even flushing). I wanted to play a bit more until I understand it and can be confident that the method works as expected, I'll do so in the next couple of days.

Thanks for the feedback to both!

porras commented:
Build failed in `linux32` but I *think* it's a random failure, unrelated to the change:

```
....................................................................................................................................
Broadcast message from root@testing-gce-d9bfa911-53cf-4bf9-b939-50f6d144dfd7
	(unknown) at 22:02 ...
The system is going down for power off NOW!
make: *** [spec] Terminated
```

[(complete log)](https://travis-ci.org/manastech/crystal/jobs/81480566)

Since I have pending changes that will trigger a new build, I'll just *“let it be... :notes:”*

jhass commented:
Yeah, already had a good laugh at that one yesterday. Travis messed up there :P

asterite commented:
LGTM :+1:

ysbaddaden commented:
@porras can you squash your commits as a single one?

jhass commented:
@ysbaddaden https://github.com/manastech/crystal/pull/1549#discussion_r40066460

porras commented:
Hey, sorry for the slow follow up, travel + illness got me away from computers :D

I've added the suggested default (completely truncate the file). I also added a call to `#flush` to avoid the problem where some write would be pending, we truncate the file and then the pending write is done (there's a test checking exactly that).

I also rebased all commits into one, on top of current master.

jhass commented:
Great, thank you!

ozra commented:
What are thoughts on windows compatibility for Crystal stdlib in the future? Afaik ftruncate is not available in many systems - will there be a fallback? (I've used nothing but LinuxOS's since 1997, but you know... the rest of the world ;-) )

porras commented:
Oh, I didn't know it wasn't supported everywhere where libc is...

I have just checked Ruby for an example and they just document the fact that is not available everywhere: http://www.rubydoc.info/stdlib/core/File%3Atruncate. Which of course doesn't mean Crystal should do the same, but for the record.

porras commented:
Which is fun because I checked that documentation to make the API the same, but somehow my eyes flew over that note 🙈

kirbyfan64 commented:
Windows has it, it's just oddly named:

https://msdn.microsoft.com/en-us/library/windows/desktop/aa365531(v=vs.85).aspx
